subsys/shell/backends/shell_mqtt.c

```diff
@@ -0,0 +1,824 @@
+static inline int sh_mqtt_context_lock(k_timeout_t timeout)
+{
+	return k_mutex_lock(&sh_mqtt->lock, timeout);
+}
+
+/* Unlock the context of the shell mqtt */
+static inline void sh_mqtt_context_unlock(void)
+{
+	k_mutex_unlock(&sh_mqtt->lock);
+}
+
+bool __weak shell_mqtt_get_devid(char *id, int id_max_len)
+{
+	uint8_t hwinfo_id[DEVICE_ID_BIN_MAX_SIZE];
+	ssize_t length;
+
+	length = hwinfo_get_device_id(hwinfo_id, DEVICE_ID_BIN_MAX_SIZE);
+	if (length <= 0) {
+		return false;
+	}
+
+	(void)memset(id, 0, id_max_len);
+	length = bin2hex(hwinfo_id, (size_t)length, id, id_max_len - 1);
+
+	return length > 0;
+}
+
+static void prepare_fds(void)
+{
+	if (sh_mqtt->mqtt_cli.transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		sh_mqtt->fds[0].fd = sh_mqtt->mqtt_cli.transport.tcp.sock;
+	}
+
+	sh_mqtt->fds[0].events = ZSOCK_POLLIN;
+	sh_mqtt->nfds = 1;
+}
+
+static void clear_fds(void)
+{
+	sh_mqtt->nfds = 0;
+}
+
+static int wait(int timeout)
+{
+	int rc = 0;
+
+	if (sh_mqtt->nfds > 0) {
+		rc = zsock_poll(sh_mqtt->fds, sh_mqtt->nfds, timeout);
+		if (rc < 0) {
+			LOG_ERR("poll error: %d", errno);
+		}
+	}
+
+	return rc;
+}
+
+/* Query IP address for the broker URL */
+static int get_mqtt_broker_addrinfo(void)
+{
+	int rc;
+
+	sh_mqtt->hints.ai_family = AF_INET;
+	sh_mqtt->hints.ai_socktype = SOCK_STREAM;
+	sh_mqtt->hints.ai_protocol = 0;
+
+	rc = zsock_getaddrinfo(CONFIG_SHELL_MQTT_SERVER_ADDR,
+			       STRINGIFY(CONFIG_SHELL_MQTT_SERVER_PORT), &sh_mqtt->hints,
+			       &sh_mqtt->haddr);
+	if (rc == 0) {
+		LOG_INF("DNS%s resolved for %s:%d", "", CONFIG_SHELL_MQTT_SERVER_ADDR,
+			CONFIG_SHELL_MQTT_SERVER_PORT);
+
+		return 0;
+	}
+
+	LOG_ERR("DNS%s resolved for %s:%d, retrying", " not", CONFIG_SHELL_MQTT_SERVER_ADDR,
+		CONFIG_SHELL_MQTT_SERVER_PORT);
+
+	return rc;
+}
+
+/* Close MQTT connection properly and cleanup socket */
+static void sh_mqtt_close_and_cleanup(void)
+{
+	/* Initialize to negative value so that the mqtt_abort case can run */
+	int rc = -1;
+
+	/* If both network & mqtt connected, mqtt_disconnect will send a
+	 * disconnection packet to the broker, it will invoke
+	 * mqtt_evt_handler:MQTT_EVT_DISCONNECT if success
+	 */
+	if ((sh_mqtt->network_state == SHELL_MQTT_NETWORK_CONNECTED) &&
+	    (sh_mqtt->transport_state == SHELL_MQTT_TRANSPORT_CONNECTED)) {
+		rc = mqtt_disconnect(&sh_mqtt->mqtt_cli);
+	}
+
+	/* If network/mqtt disconnected, or mqtt_disconnect failed, do mqtt_abort */
+	if (rc) {
+		/* mqtt_abort doesn't send disconnection packet to the broker, but it
+		 * makes sure that the MQTT connection is aborted locally and will
+		 * always invoke mqtt_evt_handler:MQTT_EVT_DISCONNECT
+		 */
+		(void)mqtt_abort(&sh_mqtt->mqtt_cli);
+	}
+
+	/* Cleanup socket */
+	clear_fds();
+}
+
+static void broker_init(void)
+{
+	struct sockaddr_in *broker4 = (struct sockaddr_in *)&sh_mqtt->broker;
+
+	broker4->sin_family = AF_INET;
+	broker4->sin_port = htons(CONFIG_SHELL_MQTT_SERVER_PORT);
+
+#if defined(CONFIG_DNS_RESOLVER)
```

Comment:
The module does `select DNS_RESOLVER` - so I think it'd be safe to assume that DNS is always enabled.

Reply:
Fixed, thanks